commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt

```diff
@@ -0,0 +1,190 @@
+package com.github.quillraven.commons.audio
+
+import com.badlogic.gdx.audio.Music
+import com.badlogic.gdx.audio.Sound
+import com.badlogic.gdx.utils.ObjectMap
+import com.badlogic.gdx.utils.Pool
+import kotlinx.coroutines.launch
+import ktx.assets.async.AssetStorage
+import ktx.async.KtxAsync
+import ktx.collections.getOrPut
+import ktx.collections.iterate
+import ktx.collections.set
+import ktx.log.debug
+import ktx.log.error
+import ktx.log.info
+import ktx.log.logger
+import kotlin.math.max
+
+/**
+ * A request for a [Sound] that is created within the [QueueAudioService.playSound] method.
+ * It contains the [filePath] and the [volume] of the request.
+ */
+private class SoundRequest : Pool.Poolable {
+  var filePath: String = ""
+  var volume = 1f
+
+  override fun reset() {
+    filePath = ""
+    volume = 1f
+  }
+}
+
+/**
+ * Implementation of [AudioService] to play one [Music] track at a time and to queue [Sound] playback
+ * to avoid multiple [Sound] instances of the same sound within a single frame.
+ *
+ * Requires an [assetStorage] to load and unload the [Sound] and [Music] resources.
+ *
+ * Use [maxSimultaneousSounds] to define the maximum number of sounds that can play concurrently within one frame.
+ *
+ * Use [maxCachedSounds] to define the size of the sound cache. [Sound] instances are cached internally for
+ * faster access of future calls. If the [maxCachedSounds] is reached then the cache gets cleared.
+ */
+class QueueAudioService(
+  private val assetStorage: AssetStorage,
+  private val maxSimultaneousSounds: Int = 16,
+  private val maxCachedSounds: Int = 50
+) : AudioService {
+
+  private val soundRequests = ObjectMap<String, SoundRequest>(maxSimultaneousSounds)
+  private val soundCache = ObjectMap<String, Sound>(maxCachedSounds)
+  private var currentMusicFilePath: String = ""
+  private var currentMusic: Music? = null
+
+  /**
+   * Adds a [SoundRequest] to the request queue for the [soundFilePath] and [volume].
+   * If the [maxSimultaneousSounds] is reached for the current frame then this request is ignored.
+   * Making a call to [update] will play all pending requests.
+   */
+  override fun playSound(soundFilePath: String, volume: Float) {
+    if (soundRequests.size >= maxSimultaneousSounds) {
+      LOG.info { "Reached maximum of '$maxSimultaneousSounds' simultaneous sounds. Request is ignored" }
```

Comment:
Might be a bit too high logging level for this, seems like `debug` would be sufficient.